flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/deltajoin/StreamingDeltaJoinOperator.java

```diff
@@ -0,0 +1,665 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** This operator supports delta join in table layer. */
+public class StreamingDeltaJoinOperator
+        extends TableAbstractCoUdfStreamOperator<
+                RowData, AsyncDeltaJoinRunner, AsyncDeltaJoinRunner>
+        implements TwoInputStreamOperator<RowData, RowData, RowData>, BoundedMultiInput {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamingDeltaJoinOperator.class);
+
+    private static final long serialVersionUID = 1L;
+
+    private static final int LEFT_INPUT_INDEX = 0;
+    private static final int RIGHT_INPUT_INDEX = 1;
+
+    private static final String STATE_NAME = "_delta_join_operator_async_wait_state_";
+
+    private final StreamRecord<RowData> leftEmptyStreamRecord;
+    private final StreamRecord<RowData> rightEmptyStreamRecord;
+
+    /** Selector to get join key from left input. */
+    private final RowDataKeySelector leftJoinKeySelector;
+
+    /** Selector to get join key from right input. */
+    private final RowDataKeySelector rightJoinKeySelector;
+
+    /** Timeout for the async collectors. */
+    private final long timeout;
+
+    /** Max number of inflight invocation. */
+    private final int capacity;
+
+    private transient boolean needDeepCopy;
+
+    /** {@link TypeSerializer} for left side inputs while making snapshots. */
+    private transient StreamElementSerializer<RowData> leftStreamElementSerializer;
+
+    /** {@link TypeSerializer} for right side inputs while making snapshots. */
+    private transient StreamElementSerializer<RowData> rightStreamElementSerializer;
+
+    private transient TimestampedCollector<RowData> timestampedCollector;
+
+    /**
+     * Recovered input stream elements backed by keyed state.
+     *
+     * <p>{@code <LeftElementRecord, RightElementRecord, Watermark, InputIndex>}.
+     */
+    private transient ListState<Tuple4<StreamElement, StreamElement, StreamElement, Integer>>
+            recoveredStreamElements;
+
+    /** Structure to control the process order of input records. */
+    private transient TableAsyncExecutionController<RowData, RowData, RowData>
+            asyncExecutionController;
+
+    /** Mailbox executor used to yield while waiting for buffers to empty. */
+    private final transient MailboxExecutor mailboxExecutor;
+
+    private final transient AtomicInteger totalInflightNum;
+
+    private final boolean[] isInputEnded;
+
+    private final transient AtomicLong asyncIOTime = new AtomicLong(Long.MIN_VALUE);
```

Comment:
It's better to move totalInflightNum and asyncIOTime at last and use `========= Metrics =========` to separate them from other fields. 

BTW, we can also assign the initial value for `totalInflightNum`.